QUT-Motorsport/QUTMS_Driverless
# Lifecycle Nodes

## What type of PR is this? (check all applicable)
Feature

## Description / Documentation
Lots of lines changed but fundamentally its a pretty simple change. Lifecycle nodes just allow other nodes to activate and deactivate it. The LifecycleNode provides several method that need to be overridden by classes that inherit it, these are on_configure, on_activate, on_deactivate, on_shutdown, on_cleanup and on_error. These correspond with states in the LifecycleNode state machine, they are called when the state is 'transitioned' into, which is initiated by the node's manager. In these state callbacks, we simply activate/deactivate publishers and subscribers, rendering the node enabled or disabled. In our case, the trackdrive_handler_node is the manager of three lifecycle nodes, a pure pursuit node, a reactive controller and a path planner. A convenience class called LifecycleServiceClient has been created to simplify the managing of LifecycleNodes. 

## To Test / Run
1. Run the sim
2. Run particle pursuit
3. Run reactive controller
4. Run ordered mid spline
5. Run trackdrive handler

The reactive controller should perform one lap, then it will be deactivated, and the trackdrive handler will activate the pure pursuit and ordered mid spline nodes. After 10 laps all nodes (except trackdrive handler) should be shutdown.

## Usability concerns or breaking changes?
Shouldn't be. We will see.
